### ISSUE TITLE
Navigation Overlay customisation via Template Parts

### ISSUE THREAD

**Issue by jasmussen** [WordPress/gutenberg#43852]
Closes https://github.com/WordPress/gutenberg/issues/39142 and https://github.com/WordPress/gutenberg/issues/38201

## Update

Following discussion in this Issue the aim is now to make it possible to define the overlay as a separate **template part**.

This will allow users to create a totally _custom_ overlay and associate that with a given Navigation block via the template part's unique identifier.

As template parts are a blank canvas, it will be possible to have a different Navigation block within the template part itself which will solve the requirement whereby some users wish to have a different menu on "mobile".

Styling of this template will be achieved via https://github.com/WordPress/gutenberg/issues/40318.

This Issue is an "epic" and will need to be further broken down into it's constituent parts.

## User Requirements

These are the high level requirements for the feature.

- As a user...
    - I should be able to customise the visual styling of the overlay that appears for a given Navigation block.
    - I should see the Navigation block within my overlay is – by default – automatically kept in sync with menu used in the parent Navigation block.
    - I should be able to display different items in my overlay Navigation than those shown in the parent Navigation.
    - I should be able to style the Navigation block used in my overlay differently than the parent Navigation block.
    - I should not be able to make my overlay inaccessible to any user.
    - I should not be able to “break” the interactive functionality of the overlay using the editor.
    - I should be able to style my overlay via Global Styles.
- As a theme creator...
    - I should be able to style my overlay via Theme JSON.
    - I should be able to provide a default navigation overlay template that will override the default one supplied by core.

## Glossary

- "Parent Navigation" - the navigation block which controls/triggers the overlay.

---

<details>
<summary>Issue update: Dec 23</summary>
To enable full customization of the contents of the Navigation overlay menu, it could become a separate block, with its own possible inner blocks, paddings, and global styles properties. To enable that, we could have the overlay button invoke a "focus mode" editing of the overlay block:

<img width="3120" alt="Editing navigation in isolation" src="https://user-images.githubusercontent.com/1204802/194289457-239c5032-52bc-4d42-b8fd-92a5bfd31231.png">

An additional affordance, for example for when the overlay only invokes on mobile, could be to add an Edit button in the Display panel of the inspector:

<div align="center">
<img width="364" alt="Navigation" src="https://user-images.githubusercontent.com/1204802/194289561-9c344b90-1ff8-4e4a-8b70-be6f8a8cc99a.png">
</div>

The Overlay could be surfaced in the List View in the same style as position: sticky; group blocks in #30121:

<img width="465" alt="List view i2" src="https://user-images.githubusercontent.com/1204802/200323295-702b1e5a-4cef-4688-a196-08f353363091.png">

However it would not be a requirement for this issue, and if helpful to the development effort, there is no need to optimize for more than a single navigation menu with a single overlay that are directly tied together with no ability to replace one overlay with another.

There are technical questions to consider ([please see this conversation](https://github.com/WordPress/gutenberg/issues/44182#issuecomment-1269560201)), it would be good to consider these in light of template parts as well, since ideally the experience of entering and existing focused mode can feel similar.

The overlay block should be stylable in Global Styles (a la `"core/navigation-overlay": {)`, so custom paddings and styles can be applied. This should also address #44182. 

[Figma](https://www.figma.com/file/vmHxwa8WhiyiP9poe2om9J/Inspector-Panels?node-id=13480%3A41897).
</details>

---

<details>
<summary>Initial proposal (Issue updated Nov 7.)</summary>

As of #43576, it is now easier to position the little navigation overlay menu icon (burger menu) so that the close icon from the overlay will perfectly overlap it. To do that, you:

* Ensure the navigation sits inside a _wide_ container block, which itself is inside a container that _inherits layout_
* Ensure that the navigation menu sits aligned to the top/right of this container and has no additional padding or margin
* The theme opts into root paddings (https://github.com/WordPress/gutenberg/pull/42085), by adding "useRootPaddingAwareAlignments": true in theme.json under settings, and paddings to go with that either in global styles or under styles > spacing.

The result is that the menu icon will sit in the exact place where the close icon will sit:

![currently](https://user-images.githubusercontent.com/1204802/188431073-3f68dd1a-72d0-4a05-b064-197dfb089693.gif)

This is working well. However, it works mainly well for themes that look good with the menu icon in the extreme top right corners. Other designs would benefit from a little vertical padding above the menu item, to optically balance, for example, a site title with the burger icon. Like so:

![allow padding customization in overlay](https://user-images.githubusercontent.com/1204802/188431384-56072965-343e-4898-b5ca-2fa967f00481.gif)

At the moment, all 4 paddings for the overlay menu are set to be the same as any root padding applied to your site. It would be nice to be able to customize these paddings further, by simply surfacing this padding control for the overlay menu, so for example the top padding could be set just a bit higher, to match any custom header padding, making things line up. 

For starters, this could be implemented in theme.json only, and a padding control for the overlay could be added to the block at a later point.

</details>

**Comment by jasmussen** [comment]
CC: @WordPress/gutenberg-design @MaggieCabrera — is this feasible in short order, you think?

**Comment by MaggieCabrera** [comment]
How would you target the overlay with just theme.json, do you think? right now we have control over the padding of the whole block, not the overlay. I don't think it's easy enough to do in time for 6.1 I'm afraid.

**Comment by jasmussen** [comment]
Thanks for looking. Good question, my initial instinct would be something like this:

```
			"core/navigation": {
				"typography": {
					"fontSize": "var(--wp--preset--font-size--medium)",
					"lineHeight": "var(--wp--preset--font-size--large)"
				}, 
				"overlay": {
					"spacing": {
						"padding": {
							"top": "80px",
							"right": "40px",
							"bottom": "80px",
							"left": "40px"
						}
					}
				}
			},
```

However if the overlay itself becomes a separate block, that would change this. 🤔

**Comment by MaggieCabrera** [comment]
yeah, I wonder about that as well, @getdave @scruffian @draganescu is that what's going to happen with the overlay in the future?

**Comment by carolinan** [comment]
Perhaps it would be something that we either
-  Select to "edit in isolation" like we edit template parts.
-  Edit in place, but hide until selected (like WooCommerce empty and filled carts). 

I think the solution could be used in other places like for the "no results" block.

**Comment by jasmussen** [comment]
I like the idea that we can edit in isolation, and I previously explored a mockup to that effect:

<img width="3000" alt="Editing navigation in isolation" src="https://user-images.githubusercontent.com/1204802/188563796-c7d09a4c-843c-4a43-b8e2-892a2b1e2ae0.png">

I think that mockup may be posted on one of the open issues. The hope with this issue, however, was finding a solution that could live in theme.json for the near future, and then be opened up for UI editing later on.

**Comment by scruffian** [comment]
The issue I see with introducing an "overlay" property in the navigation block is that it would be specific to that block. At present all the properties we have are valid for all blocks. IMO the simplest approach here is simply to make the overlay a different block.

**Comment by jasmussen** [comment]
This issue has been updated to tackle the idea of making the overlay menu a separate block that can be edited in focus mode. If you're subscribed, it could use both technical and design feedback!

**Comment by mtias** [comment]
I think it should be a template part, not a separate block.

**Comment by draganescu** [comment]
I agree w/ @mtias because I think it's a good idea:

- we keep how the block works now as a default
- if the theme or the user creates a navigation-overlay template part the block automatically uses that

Whill this template part be able to include the block that references it? 😁

**Comment by MaggieCabrera** [comment]
The point of it being a separate block is so it has different customization options: the overlay usually doesn't share the same spacing options or typography settings as the desktop navigation does. How can we accomplish that with template parts?

**Comment by Retrofitterdk** [comment]
A template part seems like a more flexible way to go. I see use cases where the navigation overlay needs to contain other elements than menus. it could be 

-  a login/logout button
- a search field

Or maybe multiple menus. Maybe the desktop site has more than one menu, and instead creating a third specific mobile menu you want the mobile menu to combine the two desktop menus in order to reduce maintenance burden for editors.

**Comment by mtias** [comment]
@MaggieCabrera with https://github.com/WordPress/gutenberg/issues/40318

**Comment by getdave** [comment]
With the addition of https://github.com/WordPress/gutenberg/issues/40318 the template part route seems like a good choice. The obvious downside would be that we'll need to find a way to making it obvious/easy for the user to _know_ that they need to do this in order to create a different overlay.

**Comment by jasmussen** [comment]
Let's move this to the dev category! @getdave feel free to update the title to reflect the template part mechanics. 🙏

**Comment by getdave** [comment]
@jasmussen I've gone ahead and updated the title and the description to reflect the "template part" route. 

However, I don't see this as being ready for dev in the sense that we don't have a clear UX for how the interactions for creating and associating template parts with a given Navigation block will work. There are lots of edge cases to be explored and I believe this will be quicker and more effective to explore in Figma than in code.

This project will have a good degree of complexity involved and in order to be as efficient as possible it would be very helpful if we could (collectively) explore this model in more detail before we proceed.

Do you agree?

**Comment by jasmussen** [comment]
Sure, can you list out a couple of flows that you'd like to see outlined better? The current mockups assume as little UI as possible, as a goal to strive for: you toggle the "Mobile" or "Always" part of the segmented control, and boom: now you have a separately editable overlay with no other action needing to be taken. Do we need user-steps in between?

**Comment by jameskoster** [comment]
Should the navigation overlay appear in the template parts list? Does it need a new semantic area so that you can have multiple overlays for different navigation menus?

**Comment by mtias** [comment]
It should appear in the template parts list; I don't think it should be a semantic area.

**Comment by jameskoster** [comment]
Just to connect a dot in case it affects implementation, this may not be the only template part that core will provide, and themes be able to override. See: https://github.com/WordPress/gutenberg/issues/44362

**Comment by getdave** [comment]
> ...you toggle the "Mobile" or "Always" part of the segmented control, and boom: now you have a separately editable overlay with no other action needing to be taken. 

I understand that you see this behaving exactly in the same way as it does now except that under the hood the overlay would be a template part rather than the exact same block as the Navigation block? Is that right?

With respect, whilst that sounds ideal, I feel there might be a few unexplored threads in that approach which if not unpicked early could lead to inefficiencies down the line. 

For example a few unexplored areas/question of the top of my head:

- how does the user use a different menu in the overlay than the one in the main Nav block? 
- how do we communicate to users the relationship between the Navigation block and the template part used for the overlay?
- what's the user experience for interacting with a Navigation block which then becomes a template part when previewing on  a smaller screen?
- as a user what's the workflow for editing my overlay menu vs the one in the Nav block?
- how do I customise the styles in the main Nav block vs the one in the overlay?
- how do I customise the overlay by adding more blocks to it (e.g. social icons)?

Ideally it would be great to see a visual UX exploration of these elements so we can unpick where the flows start to become confused before we dive too far into the technical implementation.

As it stands I'm not yet able to fully visualise all the flows and feel that if a developer jumps in here they are going to make a lot of assumptions that could lead to "dead ends". Ideally we'd anticipate those and head them off before we reach that uncomfortable point.

Moreover, I feel a visual really helps to

- crystallise ideas 
- help other contributors understand the aims - especially for those who aren't able to be as fully involved as we are.
- illustrate further potential problem areas
- highlight areas where other efforts on the project can help/inform this one

Appreciate its additional effort but I really think it would be valuable 🙇🏻‍♂️

**Comment by jasmussen** [comment]
Thanks for the elaborate response.

> how does the user use a different menu in the overlay than the one in the main Nav block?

If the template part is entirely customizable, such as an empty group you can insert anything into, then any navigation menu ite might come pre-populated with, would be editable just as if it was a second menu, no? I don't know whether this is better than a strong locked connection to the parent navigation container, just that this seems like the simplest path forward.

> how do we communicate to users the relationship between the Navigation block and the template part used for the overlay?

If the template part editing interface is opened from the main navigation block, I would think that makes the relationship between the two implicit:

<img width="324" alt="Screenshot 2022-10-24 at 12 45 48" src="https://user-images.githubusercontent.com/1204802/197509007-08ddab9b-5cef-4167-af76-235d08593f64.png">

But I think I sense your question: if the overlay is a template part, shouldn't it be possible to replace one with another, such as how you can do it for other template parts? I.e. the Replace item here:

<img width="600" alt="Screenshot 2022-10-24 at 12 46 39" src="https://user-images.githubusercontent.com/1204802/197509263-59f8b4da-9e02-43c5-b85e-9faa801feb6d.png">

In that light, I wonder if the overlay should show up in the list view? 🤔 

This is a very quick and dirty mockup, just to entertain the thought:

<img width="465" alt="List view" src="https://user-images.githubusercontent.com/1204802/197510132-b3d58fb7-1657-4dcd-a21b-835326341d08.png">

It would probably have to be locked somehow. But the principle is the same: if it's a template part, perhaps it should show up there.

> what's the user experience for interacting with a Navigation block which then becomes a template part when previewing on a smaller screen?

Ideally same as any other block or template part. And in light of that, does list view inclusion, if feasible, change the conversation?

> as a user what's the workflow for editing my overlay menu vs the one in the Nav block?

I sense there's depth to this question that is eluding me, so please let me know what I'm missing if anything. But for the initial proposal, you'd edit in two ways, first by clicking the menu icon in the canvas:

<img width="570" alt="Screenshot 2022-10-24 at 13 02 59" src="https://user-images.githubusercontent.com/1204802/197511960-2e5a6353-8a73-4a80-b022-9a6d1ef14ede.png">

And if that menu icon isn't visible, perhaps because you're at the desktop breakpoint and have the overlay set to only show when on mobile, then from the inspector:
<img width="520" alt="Screenshot 2022-10-24 at 13 03 05" src="https://user-images.githubusercontent.com/1204802/197512049-25652e38-1171-43e9-879d-77a6033f5bc8.png">

I suppose if we wanted to add the overlay as accessible in the list view, you could select that block to open it as well.

> how do I customise the styles in the main Nav block vs the one in the overlay?

Select the nav block, and you have the nav block inspector and toolbar available.

In the initial proposal: the navigation overlay as a block, well you'd enter focus mode for just that block and have access to its inspector. I think this can remain mostly the same with it as a template part:
<img width="798" alt="Screenshot 2022-10-24 at 13 06 59" src="https://user-images.githubusercontent.com/1204802/197512472-6abe7301-481c-45db-8f5d-948dfbb8fe98.png">

<img width="1392" alt="Screenshot 2022-10-24 at 13 07 05" src="https://user-images.githubusercontent.com/1204802/197512487-a6d0190e-ec93-445c-a71b-6a44f3bce937.png">

Only, the "Edit" from the block toolbar would be the "Edit" in the inspector, or just clicking the burger menu button if available.

> how do I customise the overlay by adding more blocks to it (e.g. social icons)?

I'm thinking of it just as a group. You can add anything inside.

**Comment by talldan** [comment]
> In that light, I wonder if the overlay should show up in the list view? 🤔

Could there be some overlap with the design explorations over on fixed/sticky blocks - https://github.com/WordPress/gutenberg/issues/30121#issuecomment-1286816605. The overlay is technically a fixed element.

**Comment by jasmussen** [comment]
Nice, yes indeed! No reason it couldn't show up in the same style as @javierarce designed there 👍 👍

**Comment by jameskoster** [comment]
> If the template part editing interface is opened from the main navigation block, I would think that makes the relationship between the two implicit

Agreed. It would be really strange if you could instruct a navigation menu to _always_ use the overlay, then insert a _different_ navigation menu in the overlay. What if the menu in the overlay also uses an overlay 🫠? 

> if the overlay is a template part, shouldn't it be possible to replace one with another, such as how you can do it for other template parts?

If the overlay and the navigation menu are intrinsically linked, then replacing doesn't make much sense, unless I am missing something (quite likely)? What we might have to consider is dynamic creation of overlays for each navigation menu. IE `primary-menu-overlay`, `secondary-menu-overlay`.

The intrinsic link is unique, and we may need a treatment in the UI to communicate this "If you add a menu item in the overlay, it will be added to the menu as well".

**Comment by jasmussen** [comment]
I updated the issue with a mockup for surfacing the overlay in the list view, and added a note that we can start the effort with just a single navigation being tied to a single overlay, with no need to worry about replacing one with another. Let's keep things as simple as we can.

**Comment by carolinan** [comment]
_Thinking out loud,_ Should the navigation block itself register the template part if the template part file does not already exist in the theme?
I think that will be necessary if we want to allow different navigations to have different overlays?

**Comment by carolinan** [comment]
What do we need to do to prevent navigation blocks with overlays _inside navigation block overlays?_
Any template part can have a navigation block inside it, so we can't simply hide the controls if the navigation block is inside any template part.

**Comment by carolinan** [comment]
Once the template part is opened in isolation, should the back button be available?
I think for usability, it needs to be available, but how do we pass the correct `location` from the origin _block_?
**Edit:** This is in context of the block editor where we don't have access to the site editor routes.

**Comment by carolinan** [comment]
One more question, how should the close button be displayed?
- The editor and front should match.
- The button must not be removable. 
- Should it be editable in the canvas?

**Comment by annezazu** [comment]
Tagging in @WordPress/gutenberg-design for thoughts here :)

**Comment by carolinan** [comment]
This may be obvious to others but I have been so focused on block themes:
How do we solve this for classic themes that do not have template parts?

**Comment by jameskoster** [comment]
These are all excellent questions and speak to the difficulty of this issue. 

> Should the navigation block itself register the template part if the template part file does not already exist in the theme?

Yes that seems necessary, otherwise the Overlay option becomes theme dependent. I don't know if it's the Navigation block that registers it, or whether it's built in like we've begun to discuss for the Post Template block in Query Loops (https://github.com/WordPress/gutenberg/issues/44362).

> What do we need to do to prevent navigation blocks with overlays inside navigation block overlays?

This is one of the most difficult elements. I agree we don't want nested overlays, but do we want to fully prohibit the insertion of Navigation blocks here? Ultimately the full-featured Navigation block is _probably_ overkill for this purpose, especially because syncing is inherent by virtue of being a template part already. It could begin to get very confusing if your navigation menus inside overlays begin to mix with 'standard' ones.

In any case, the Overlay template part will likely need to be marked somehow in order to invoke some of its unique properties:

1. It does not appear in Inserters or replace flows.
2. It cannot be deleted.
3. It contains a Close Button block which should not be removed (related: https://github.com/WordPress/gutenberg/issues/40618).
4. Potentially prohibit the insertion of nested Navigation blocks as described above.

> Once the template part is opened in isolation, should the back button be available?

I would expect this behaviour to match other template parts. IE if you open it from the management list there's no back button, but if you open it while editing a template then there is. 

---

One other big question I don't think we discussed yet: should it be possible to create multiple overlays? If so how does one link Navigation block A to Overlay B?

**Comment by carolinan** [comment]
> It could begin to get very confusing if your navigation menus inside overlays begin to mix with 'standard' ones.

Yes. Especially if there is only one _menu_ and this menu is reused when adding the new navigation block inside the overlay.
Yet if the content is used for navigation, it needs to use the correct HTML element.

**Comment by carolinan** [comment]
Maybe the inserter and replacement exceptions can be made by using a new template part area called overlay.

**Comment by jameskoster** [comment]
It makes sense to me, but I don't think @mtias was keen. It might be good to get some additional context as to why.

If we _did_ create a new area, perhaps "Modal" would be a good name, and better position us for general modal creation in the future? It's quite easy to imagine an augmentation of the link control UI that surfaces template parts assigned to the `modal` area as destination options. Crude mockup:

<img width="421" alt="Screenshot 2023-01-13 at 17 28 30" src="https://user-images.githubusercontent.com/846565/212381867-434b33e1-1a1e-4da7-b6da-2c842ad115bf.png">

**Comment by jameskoster** [comment]
Alternatively, we might create a new primary template type; `modal.html`. If it exists, then the link control UI would surface an "Open in modal" toggle when a link to an internal post / page exists: 

<img width="502" alt="Screenshot 2023-01-14 at 16 03 50" src="https://user-images.githubusercontent.com/846565/212481629-1ae52a73-aec1-4bdb-8ceb-965622b9b864.png">

Clicking the link on the frontend would open the associated document using the `modal.html` template. I appreciate at this point I've veered away from Navigation despite the tangential relevance. But I figured it was worth rounding out the idea since there's quite a lot of overlap here... probably worth investigating more in a separate issue, I digress.

---

To bring things back on topic, perhaps we begin with a very simple implementation here:

1. For each navigation menu a corresponding `overlay-menu-name.html` template part exists and is populated with the same link items by default.
2. We can (hopefully) use the slugs to apply special rules (e.g. doesn't appear in Inserters), no nested Navigation blocks, etc.
3. When a Navigation menu displays the hamburger icon, it will always open the associated template part on click.

The only thing missing there is how to handle scenarios where no navigation menus exist, but that doesn't seem insurmountable.

**Comment by annezazu** [comment]
Noting that this has been deprioritized for the WordPress 6.2 release.

**Comment by carolinan** [comment]
This needs a design refresh and more exploration because of the changes to the navigation block list view and menu selection.

**Comment by carolinan** [comment]
I still think this is one of the bigger questions:
How do we solve this for classic themes that do not have template parts?

**Comment by getdave** [comment]
Related https://github.com/WordPress/gutenberg/issues/38201

**Comment by priethor** [comment]
> How do we solve this for classic themes that do not have template parts?

Since this feature would allow for completely _new_ designs, I don't think we need to solve this for classic themes that do not have template parts. Even though classic themes could use block template parts, the goal is not to have one-to-one feature parity between classic and block themes but to provide migration paths and an easier transition to the new paradigm.

**Comment by annezazu** [comment]
Removing this from 6.4's board with four weeks before beta 1. Let's also make sure the design is refreshed since it was last updated last November and a lot has happened since then.

**Comment by ResponsiveFlow** [comment]
> Removing this from 6.4's board with four weeks before beta 1.

Is it due to the fact that there’s not enough time to implement it for 6.4?
will it be included in 6.5?

**Comment by annezazu** [comment]
Correct, we don't have enough time and this isn't fleshed out enough to move forward on I'm afraid. The 6.4 cycle has been substantially shorter than other cycles so it was a tough but necessary call to make. Ahead of each release, issues like this will be evaluated once more. I can't speak specifically to the 6.5 roadmap just yet though.

**Comment by getdave** [comment]
I think that's a good move. Note that removing from 6.4 board doesn't mean it's not something contributors see as a priority. 

However as Anne noted, there is limited time in _this_ release window and contributors have been focusing on consolidating existing features for this release.

**Comment by mtias** [comment]
Something related to this issue is how we want to organize these parts / patterns in the library. I'd love to have an example in core of how plugins should also provide their own—registration, editing flow, discovery.

**Comment by richtabor** [comment]
Did some digging on this: 

The navigation block uses an overlay template part, provided by core—or by a theme. Let’s name it `navigation-overlay.html` for the sake of this walkthrough. Out of the box, the core `navigation-overlay.html` template part contains a stack block with a navigation block inside of it.

When I click “Edit” on my _first_ navigation block, it will modify the template part and save those modifications in the database as usual. I see the template part in isolated/focus mode. At this point I can edit the contents of this template part, including adding buttons, social links, or even assigning a different menu within the overlay's navigation block. 

However, if I then click “edit” on a _second_ navigaiton block, this is no longer the standard template part modification flow, because we already have a modified version of `navigation-overlay.html` from the first Nav block. 

At this point, the `navigation-overlay.html` template part (on the filesystem) is used as the source to create a new `wp_template_part` post in the database, but the connection is severed with the original template part. This is because we need unique slugs and template parts. 

This would mean that themes could provide a unique, and customizable, mobile navigation overlay, and you could have different overlays for each navigation. 

This method is being explored in https://github.com/WordPress/gutenberg/pull/55657. 

Props: @scruffian @getdave @MaggieCabrera @draganescu

**Comment by draganescu** [comment]
Here is a visual representation of what should happen, in my understanding (click to expand for details):

![Untitled-2023-11-02-1054](https://github.com/WordPress/gutenberg/assets/107534/f074ef48-c481-4996-833a-b1e87b2e70be)


### Details:

- the navigation **block** has a default hardcoded overlay
  - this default is an overlay block with one navigation block inside
  - the overlay block is like a group block in terms of supports, but only exists so that themes can style the overlay colors, fonts and spacing of the navigation overlay
- for every navigation, new or existing, editing the overlay will create a wp_navigation_overlay post based on the default overlay content
- in a navigation block, changing the navigation _also changes the overlay_
- two navigations cannot share the same overlay
- styling of the overlay can be globally preserved by applying overlay block styles as global
- whenever an overlay is edited for the 1st time the ref of the navigation block in the overlay is set to be the same as the ref of the navigation block whose overlay is edited
  - once the overlay is edited (or simply a custom one is created) the ref of the navigation block can be changed and it will not be changed back

Alternatively:

- we could do without an overlay block
- themes can provide a template part which always has to have the same name `parts/navigation_overlay.html`
- the navigation block starts either from the theme template part or from its default makrup
- if the theme provides an overlay template part the user always starts from the template part, which can also be independently customized and the customized template part serves as the new base.

**Comment by luisherranz** [comment]
Happy to see some movement here 🙂

Feel free to ping @WordPress/interactivity-api for anything that needs to be changed in the frontend implementation.

**Comment by scruffian** [comment]
With respect to styles there are a few challenges we need to consider:

1. Navigation blocks already provide tools to style the colors in the overlay.
2. These styles are tied to the submenu - they are the same.
3. For backwards compatibility we cannot break these colors on existing blocks.

In addition to these challenges, we need to preserve the current overlay behaviour to avoid making changes on users sites. I think therefore that if a navigation block has a custom overlay we should update the labels on the color settings to only apply to submenus and not to the overlay.

**Comment by scruffian** [comment]
I also think we need to do something to preserve the relationship between a navigation block and the navigation inside the overlay - these two should always be kept in sync unless the navigation inside the overlay is explicitly edited.

**Comment by getdave** [comment]
I've updated this Issue's description with a _proposed_ set of User Requirements. I expect these to act as a north star for the work on this feature. 

I think it's important that we don't loose sight of user goals/expectations.

**Comment by richtabor** [comment]
Composites of expected flows: 

**Simple:**
User one has a header navigation menu, clicks on "Edit" to edit the overlay, changes the background to black, sets alignment to center and sets the font size to large. Done. The navigation items are the same, as the refs match. When I make changes to one menu, those changes are reflected on the other. 

**Intermediate:**
User two has a header navigation menu, clicks on "Edit" to edit the overlay. This user wants to have different items in the overlay navigation. They choose a different menu in the Inspector; the menus are no longer in sync. 

**Advanced:**
User three has two header navigation menus, used throughout the site on different templates. This user clicks "Edit" to edit the overlay, changes the styles of the overlay and adds additional blocks to the overlay. This user wants to use the same customized overlay consistently throughout the site. This user needs to be able to select an existing overlay template part. (to do later)

Thoughts?

**Comment by getdave** [comment]
@richtabor These seems to align nicely with the existing user stories although the `Advanced` part has some elements that are not captured. I wonder if the User Stories can be adapted to incorporate your `Simple, Intermediate, Advanced` scenarios?

**Comment by richtabor** [comment]
Hey team, 

I summarized much of the conversations, concepts, and feedback into a [prototype](https://www.figma.com/proto/oEm1ck3f5gfGYyOR1Fs71T/Gutenberg-Ideas?page-id=1153%3A15251&type=design&node-id=1197-9781&viewport=-1641%2C6535%2C0.62&t=2vE6HlfPAMj5G7nq-1&scaling=min-zoom&starting-point-node-id=1197%3A9781&show-proto-sidebar=1&mode=design), and make a video walkthrough. 

There are a few rough edges to iron out yet, but I think we're on the right path. Not shown is editing overlays in the site editor > patterns view, and a next iteration where you may potentially change the ref of the menu within the overlay.

Thoughts?

https://github.com/WordPress/gutenberg/assets/1813435/49a27152-0a12-4472-a4a9-2bcc6b32c0ef

**Comment by paaljoachim** [comment]
@richtabor 
Rich. Wow this is awesome! Adding a prototype very much helps give the visual feel of what it might look like!

Clicking Edit to enter the Edit overlay mode -> Shows Navigation overlay in the top which is very helpful.

In relation to editing the menu elements specific to the overlay. Can we add a bigger font or change the text color which will only impact the overlay and not the regular menu?

**Comment by MaggieCabrera** [comment]
> In relation to editing the menu elements specific to the overlay. Can we add a bigger font or change the text color which will only impact the overlay and not the regular menu?

Yes, the goal is that the styles of the nav block within the overlay can be totally different than outside of it, included justification/alignment and any other styling controls that you know have

**Comment by paaljoachim** [comment]
Thank you for the feedback Maggie!

**Comment by getdave** [comment]
@richtabor This aligns nicely with the current direction of this PR 👍

**Comment by getdave** [comment]
# Update

Now contributors are back at the start of the new year, I felt now was a good time to revisit this effort and provide an update on status.

We have [a PR which implements the proposal to utilise template parts allow for editing the Navigation block's Overlay](https://github.com/WordPress/gutenberg/pull/55657).

During the course of working on that PR it has become apparent that there are a number of different UX routes that could be taken in regard to determining how the correct template part is chosen and also the various scenarios that arise when making changes to the contents of the Overlay.

In short, it's my view that from a UX perspective this project is _much_ more complex than anyone first appreciated. In addition, this PR is currently only a prototype and is far from being a refined experience. There are a number of edges to the problem that need to be explored and resolved and the amount of effort required to resolve these is unknown.

As a result, I'm feeling ever less confident that we can land this PR for the 6.5 release. Even if we were to settle on a UX and solve all the technical issues in the few weeks left before Beta 1, that leaves very little time for user testing and feedback before it lands in a major WordPress release.

Therefore, I think we ought to _consider_ removing the Navigation Overlay feature from the roadmap for 6.5 and instead target it for a future release. We can however, continue to work on the feature during the 6.5 _cycle_ and aim for its inclusion within the Gutenberg _Plugin_.

As the feature is opt-in, it should be possible to achieve this. 

We will still continue to work on the feature during the 6.5 cycle but without the pressure involved in aiming for it to be included in the release. This will give contributors more breathing room to have the opportunity to receive feedback and refine the implementation.

Do we agree that, given the issues outlined above, it would be premature to target this for 6.5 and we should look to refine the feature prior to inclusion in a major release?

**Comment by scruffian** [comment]
> Do we agree that, given the issues outlined above, it would be premature to target this for 6.5 and we should look to refine the feature prior to inclusion in a major release?

Absolutely

**Comment by draganescu** [comment]
> we should look to refine the feature prior to inclusion in a major release?

Yes!

**Comment by annezazu** [comment]
Moving this to punted to 6.6 due to the feedback and discussion here. Thanks for the clear communication. Noted in the 6.5 release leads channel for good measure.

**Comment by getdave** [comment]
Noting here that I think it will be very difficult to merge this PR and target this PR at the Gutenberg Plugin only (i.e. keep the change in the Plugin and not allow it into a release).

Why? Because then we'd be maintaining two versions of the block. 

1. Plugin version - different JS, PHP and CSS. Possibly branching.
2. Core version - some standard version without the above changes.

This is branching and is going to introduce more maintenance overhead into an already complex block.

Instead I suggest we investigate whether it's possible to place it behind a _flag_. That way we can ship the **same code** to Plugin and Core and it's clear where the branching logic occurs.

**Comment by paaljoachim** [comment]
Feedback from a client in relation to the hamburger/overlay menu as it is today.
Default overlay menu as seen at: https://www.permakultur.no/ when Justify Items right has been selected for the Nav.

https://github.com/WordPress/gutenberg/assets/5323259/35673d5b-7b38-4fcc-961a-1ed50a23cbf4

Items are crammed into the right side. Needs padding. Align left - center - right. 
Top item and sub levels are not clear. Needs more clarity in relation to top item - sub levels 1 - sub levels 2 etc.
(Changing to justify items center gave a better grouping.)
The client actually suggested to have a drop down of sub level items inside the overlay.
 
Btw going to the site I changed Justify items right to Justify items center as it looked a lot nicer.

EDIT:
I forgot to mention that I turned off the CSS I had added which added padding on the right side of the menu. The top level with sub levels did not look any nicer, but atleast it added some space on the right side.

EDIT 2: Padding right existed on another Full Site Editing web site using the default theme Twenty Twenty Four.

So the main point goes back into having top level with sub levels items easier to notice.

**Comment by MaggieCabrera** [comment]
> Btw going to the site I changed Justify items right to Justify items center as it looked a lot nicer.

The lack of padding is probably a theme issue. You need to add a global horizontal padding for it (and if it's 0, make sure it's not a unitless zero)

**Comment by scruffian** [comment]
I've been thinking more about this and I think we should explore what it would take to allow a link/button to open a template part in an overlay, as a first step.

**Comment by colorful-tones** [comment]
Hi folks,
We are only one week away from the Beta 1 cut-off date for [WordPress 6.6](https://make.wordpress.org/core/6-6/). This issue hasn’t seen any movement in a while, so we (the editor triage leads of the 6.6 release) have decided to remove it from the [WordPress 6.6 Editor Tasks project board](https://github.com/orgs/WordPress/projects/179/views/1).

**Comment by davewhitley** [comment]
Hi all, here is a plugin that solves the problem in a pretty elegant, simple way. Something I think we could do in core feasibly:

https://youtu.be/Z12kUMtLODw?si=L4JFPILVpXuV_ovm&t=284

(the video is about mega menus, but that is irrelevant)

As discussed, the mobile menu would be a separate template, and you can choose the template in the sidebar.

<img width="309" height="415" alt="Image" src="https://github.com/user-attachments/assets/88d279c7-10b7-4d81-8826-3ace2313b8b0" />

**Comment by karmatosed** [comment]
Perhaps for now this more needs feedback around design.

**Comment by getdave** [comment]
This Issue is now being addressed under https://github.com/WordPress/gutenberg/issues/73084
